MIDI-36: Build model training and evaluation pipeline

JasiekKaczmarczyk opened the pull request:
[Build model training and evaluation pipeline](https://sacrebleu.atlassian.net/browse/MIDI-36)

Implementation of Unet architecture for reverse diffusion.

Here is example test plot to check if the model runs:
![reverse_diffusion_not_trained](https://github.com/Nospoko/ecg-diffusion/assets/94491798/3bf7b64f-daa7-4400-84c8-c3a62d655b84)

Commit:
MIDI-35: Forward diffusion process

Commit:
MIDI-35: Forward diffusion process

Commit:
MIDI-35: Forward diffusion process

JasiekKaczmarczyk committed:
Delete screenshots directory

Commit:
MIDI-35: Forward diffusion process

Commit:
MIDI-36: Reverse diffusion

JasiekKaczmarczyk committed:
Merge branch 'master' into MIDI-36/training-model

Commit:
MIDI-36: Training loop

JasiekKaczmarczyk commented:
I implemented training loop and some initial test for model. Just like previously I started with training model on single batch of size 16 to find if I will manage to overfit it.

Here is batch used for training:
![training_data_single_batch](https://github.com/Nospoko/ecg-diffusion/assets/94491798/deb6a1ce-0e68-4ad8-93a3-f935075ab508)

And here is result of model trained for 1000 steps on this batch:
![model_trained_single_batch](https://github.com/Nospoko/ecg-diffusion/assets/94491798/1955f12c-5ee4-4aef-ac5d-7aa219839086)

Model doesn't generate the same samples (probably due to short training), but at least it found right range of values for amplitude. One thing that is weird is that in each of the synthetic signals there are some artifacts with high amplitude at the beginning and end of signal.

Training run on wandb: https://wandb.ai/jasiekkaczmarczyk/ecg-diffusion/runs/vszoirto?workspace=user-jasiekkaczmarczyk

roszcz commented:
Can you try the sample generation, but starting from the signal + noise, instead of pure noise. Here's what I mean:

```python
for ecg_sample in batch:
  # I'm assuming that right now it's:
  # model_input = torch.randn_like(ecg_sample)

  # Use partially distorted ECG to see if the model can clean it up
  model_input = diffusion(ecg_sample, step=63)
  out = model(model_input)
  plot(out)
```

Let me know if this makes sense! 

For the edge amplitude, I think it's some kind of conv artifacts. Maybe it will be possible to get rid of those with padding, but maybe we'll need to investigate in more detail. Not directly related, but here's a great article about conv artifacts: https://distill.pub/2016/deconv-checkerboard/

roszcz commented:
@JasiekKaczmarczyk Can you also set up metrics for validation dataset?

JasiekKaczmarczyk commented:
> Can you try the sample generation, but starting from the signal + noise, instead of pure noise. Here's what I mean:
> 
> ```python
> for ecg_sample in batch:
>   # I'm assuming that right now it's:
>   # model_input = torch.randn_like(ecg_sample)
> 
>   # Use partially distorted ECG to see if the model can clean it up
>   model_input = diffusion(ecg_sample, step=63)
>   out = model(model_input)
>   plot(out)
> ```
> 
> Let me know if this makes sense!
> 
> For the edge amplitude, I think it's some kind of conv artifacts. Maybe it will be possible to get rid of those with padding, but maybe we'll need to investigate in more detail. Not directly related, but here's a great article about conv artifacts: https://distill.pub/2016/deconv-checkerboard/

Ok, I'll do it when I'll train model on whole dataset, on current model trained on single batch it probably wouldn't work.

Commit:
MIDI-36: Updated training with validation

SamuelJanas commented on tests/reverse_diffusion_test.py:
```diff
@@ -18,14 +16,7 @@
-        kernel_size=7,
-        resnet_block_groups=4
-    )
+    model = Unet(in_channels=2, out_channels=2, dim=32, dim_mults=(1, 2, 4), kernel_size=7, resnet_block_groups=4)
```
> You should use [magic trailing commas](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#the-magic-trailing-comma) to let black know that you want this split.

It's a suggestion I got from @roszcz, was helpful for those situations. In short, adding a comma at the end of `resnet_block_groups=4` would result in black keeping it not collapsed!

Commit:
MIDI-36: Updated test script

JasiekKaczmarczyk commented:
I trained model for 1 epoch and the results are promising.

Here are some samples generated by sampling pure noise and then reversing the process.
![model_trained_1](https://github.com/Nospoko/ecg-diffusion/assets/94491798/6190ebf4-3ee3-412c-a4a0-9a5d1df91682)

And here are results of samples from validation set diffused a bit with timestep 63 (full diffusion is with timestep 255) and then denoised back to original:
![model_trained_denoising_val_set](https://github.com/Nospoko/ecg-diffusion/assets/94491798/5b5d6bd7-68c1-4b55-acca-51ca815fcee3)

Loss for each of signals above:
Loss for signal 0: 0.011288120411336422
Loss for signal 1: 0.008314604870975018
Loss for signal 2: 0.005568422842770815
Loss for signal 3: 0.005224710330367088

Commit:
MIDI-36: Updated test script

JasiekKaczmarczyk commented:
I added checkpoint to huggingface and updated test script to use it instead of local checkpoint, so you all can test it. Just run ```python tests/test_sampling.py```.

roszcz commented:
Nice! Can you also measure time of inference per a single sample? What's the number of parameters in this model?

JasiekKaczmarczyk commented:
> Nice! Can you also measure time of inference per a single sample? What's the number of parameters in this model?

Ok, I'll do that and update README. This model has about 2-3 milion parameters.

Commit:
MIDI-36: Updated README

JasiekKaczmarczyk commented:
> Nice! Can you also measure time of inference per a single sample? What's the number of parameters in this model?

I ran inference several times and for gpu it takes about 1.5-2.5 seconds for single sample and for cpu 7-8 seconds.

roszcz commented on train.py:
```diff
@@ -0,0 +1,200 @@
+
+            val_step_count += 1
+
+            if (batch_idx + 1) % cfg.logger.log_every_n_steps == 0:
+                # log metrics
+                wandb.log(metrics, step=val_step_count)
```
For validation it would be better to get the metrics averaged over the whole dataset, rather than for just some batches.

I think you can also use the `train_step_count` here, so the values will be aligned with the train stats. Alternatively you can also do `step=epoch`. (Assuming you'll calculate metrics for the whole set)

roszcz commented on train.py:
```diff
@@ -0,0 +1,200 @@
+
+        for batch_idx, batch in val_loop:
+            # metrics returns loss and additional metrics if specified in step function
+            metrics = step(unet, forward_diffusion, batch, device, split="val")
```
I'm not sure if it worth it, but you can try using `with torch.no_grad()` context here to improve performance :thinking:

JasiekKaczmarczyk replied:
Ok, I'll look into it

roszcz commented on train.py:
```diff
@@ -0,0 +1,200 @@
+    # get loss value for batch
+    loss = F.mse_loss(predicted_noise, added_noise)
+
+    metrics = {f"{split}/loss": loss}
+
+    return metrics
```
```suggestion
    return loss
```

I think it would be better to separate the metrics management from the training steps. 

Also, I think storing tensors with gradients like that creates a risk of "dangling graphs" and memory leaks - which I think I'm running into during the validation step here

JasiekKaczmarczyk replied:
I think it would be better to return both loss (with grad) and metrics (without grads). When doing it this way training loop is more modular and will work when doing something else, only thing you need to change then is step function.

roszcz replied:
If this step has to return more than just the loss we can have a `dataclass` full of metrics as an output, but I wouldn't spend too much time planning modularity we don't know we need right now :)

For me, the main suggestion here is to remove `split` from this function completely, as it's mixing the model logic with performance tracking logic, which I think should be separate.

roszcz commented on train.py:
```diff
@@ -0,0 +1,200 @@
+    return train_dataloader, val_dataloader, test_dataloader
+
+
+def step(
```
```suggestion
def forward_step(
```

This is mostly for my personal convenience - I'm jumping across the file a lot with grep like tools (vim), and the word "step" has many occurrences not related to this function, which causes me to get lost quickly :)

roszcz commented on tests/test_sampling.py:
```diff
@@ -0,0 +1,126 @@
+import torch
+import seaborn as sns
```
```
echo 'seaborn' >> requirements.txt
```
:)

roszcz commented on tests/test_sampling.py:
```diff
@@ -0,0 +1,126 @@
+    ).to(cfg.train.device)
+
+    model.load_state_dict(checkpoint["model"])
+    forward_diffusion.load_state_dict(checkpoint["forward_diffusion"])
```
Looks to me the diffusion class does not actually have any trainable state, do we need to track this?

JasiekKaczmarczyk replied:
Yeah, because it has buffers (betas, alphas and so on) that are needed for sampling.

roszcz replied:
As far as I can tell these buffers store constant, pre-calculated arrays - would it be possible to do:

```python
slef.sqrt_recip_alphas = sqrt_recip_alphas

# Instead of
self.register_buffer("sqrt_recip_alphas", sqrt_recip_alphas)
```
Or am I missing something?

JasiekKaczmarczyk replied:
It wouldn't work because when ```model.to("cuda")``` is specified this ```self.sqrt_recip_alphas = sqrt_recip_alphas``` doesn't get cast to gpu. You need to register buffer to make it work.

roszcz commented on sample.py:
```diff
@@ -0,0 +1,63 @@
+        Sample ECG signal
+
+        Args:
+            x (torch.Tensor): signal to denoise
+            timesteps (int, optional): timestep from which denoising starts if None denoising starts from forward_diffusion end timestep. Defaults to None.
+            intermediate_outputs (bool, optional): if True returns dict of intermediate outputs with timestep as key. Defaults to False.
```
This is again about my personal convenience, but I work on a small monitor, and lines >120 characters long get broken and mess with visual indentation :)

I'll try to enforce setting this `flake8 --max-line-length=120` in the pre-commit hooks across the projects.

roszcz commented:
Looks like the memory consumption increases during validation - I think this shouldn't happen (unless batch size is different for validation). Does this happen on your machine as well?

<img width="666" alt="image" src="https://github.com/Nospoko/ecg-diffusion/assets/8056825/1ebfebbc-7c91-4d8f-9460-4fb589e863df">

JasiekKaczmarczyk commented:
> Looks like the memory consumption increases during validation - I think this shouldn't happen (unless batch size is different for validation). Does this happen on your machine as well?
> <img alt="image" width="666" src="https://user-images.githubusercontent.com/8056825/261937588-1ebfebbc-7c91-4d8f-9460-4fb589e863df.png">

Yeah, for me it is the same.

Commit:
MIDI-36: Minor changes

SamuelJanas commented:
I don't know if it's an reproducible error but running `python3 tests/test_sampling.py` gives me an error:
```bash Traceback (most recent call last):
Traceback (most recent call last):
  File "test_sampling.py", line 8, in <module>
    from sample import Generator
ModuleNotFoundError: No module named 'sample'
```
In my IDE it's clearly finding `sample.py` but I can't seem to run your code. I apologize in advance if the problem is on my side.

WojciechMat commented:
> > Looks like the memory consumption increases during validation - I think this shouldn't happen (unless batch size is different for validation). Does this happen on your machine as well?
> > <img alt="image" width="666" src="https://user-images.githubusercontent.com/8056825/261937588-1ebfebbc-7c91-4d8f-9460-4fb589e863df.png">
> 
> Yeah, for me it is the same.

I think you might want to delete loss and batch from memory by calling del loss and del batch after each iteration. I also use torch.cuda.empty_cache() after training.

WojciechMat commented:
> I don't know if it's an reproducible error but running `python3 tests/test_sampling.py` gives me an error:
> 
> ```shell
> Traceback (most recent call last):
>   File "test_sampling.py", line 8, in <module>
>     from sample import Generator
> ModuleNotFoundError: No module named 'sample'
> ```
> 
> In my IDE it's clearly finding `sample.py` but I can't seem to run your code. I apologize in advance if the problem is on 

You may put script you want to run outside any directory (i.e. Where the sample.py file is located) and try to run it from there. (You might need to specify directories in other imports) (I know this is not a perfect solution when it comes to project architecture but it should work)
Running 
```
python3 tests/test_sampling.py
```
From parent directory might also work maybe?
